test/satellite/process.tags.test.ts

```diff
@@ -0,0 +1,790 @@
+  await runMigrations()
+  await satellite._setAuthState()
+
+  const clientId = satellite['_authState']['clientId']
+
+  // Local INSERT
+  const stmts1 = {
+    sql: `INSERT INTO parent (id, value, other) VALUES (?, ?, ?)`,
+    args: ['1', 'local', null],
+  }
+  await adapter.runInTransaction(stmts1)
+  const txDate1 = await satellite._performSnapshot()
+
+  const localEntries1 = await satellite._getEntries()
+  const shadowEntry1 = await satellite._getOplogShadowEntry(localEntries1[0])
+
+  // shadow tag is time of snapshot
+  const tag1 = genEncodedTags(clientId, [txDate1])
+  t.is(tag1, shadowEntry1[0].tags)
+  // clearTag is empty
+  t.like(localEntries1[0], {
+    clearTags: JSON.stringify([]),
+    timestamp: txDate1.toISOString(),
+  })
+
+  // Local DELETE
+  const stmts2 = {
+    sql: `DELETE FROM parent WHERE id=?`,
+    args: ['1'],
+  }
+  await adapter.runInTransaction(stmts2)
+  const txDate2 = await satellite._performSnapshot()
+
+  const localEntries2 = await satellite._getEntries()
+  const shadowEntry2 = await satellite._getOplogShadowEntry(localEntries2[1])
+
+  // shadowTag is empty
+  t.is(0, shadowEntry2.length)
+  // clearTags contains previous shadowTag
+  t.like(localEntries2[1], {
+    clearTags: tag1,
+    timestamp: txDate2.toISOString(),
+  })
+
+  // Local INSERT
+  const stmts3 = {
+    sql: `INSERT INTO parent (id, value, other) VALUES (?, ?, ?)`,
+    args: ['1', 'local', null],
+  }
+  await adapter.runInTransaction(stmts3)
+  const txDate3 = await satellite._performSnapshot()
+
+  const localEntries3 = await satellite._getEntries()
+  const shadowEntry3 = await satellite._getOplogShadowEntry(localEntries3[1])
+
+  const tag3 = genEncodedTags(clientId, [txDate3])
+  // shadow tag is tag3
+  t.is(tag3, shadowEntry3[0].tags)
+  // clearTags is empty after a DELETE
+  t.like(localEntries3[2], {
+    clearTags: JSON.stringify([]),
+    timestamp: txDate3.toISOString(),
+  })
+
+  // deliver first transaction back and fake ack (to remove entry from oplog)
+  const oplog = satellite.opts.oplogTable.tablename.toString()
+  const stmts4 = {
+    sql: ` DELETE FROM ${oplog} WHERE timestamp = ?`,
+    args: [txDate1.toISOString()],
+  }
+  await adapter.runInTransaction(stmts4)
+  t.is(2, (await satellite._getEntries()).length)
+
+  // apply incomig operation (local operation ack)
+  const ackEntry = generateRemoteOplogEntry(
+    tableInfo,
+    'main',
+    'parent',
+    OPTYPES.insert,
+    txDate1,
+    tag1,
+    {
+      id: 1,
+      value: 'local',
+      other: null,
+    },
+    undefined
+  )
+
+  await satellite._apply([ackEntry], clientId)
+  let shadow = await satellite._getOplogShadowEntry()
+
+  t.like(
+    shadow[0],
+    {
+      tags: genEncodedTags(clientId, [txDate3]),
+    },
+    'error: tag1 was reintroduced after merging acked operation'
+  )
+})
+
+// Problem of previous test does not occur with just two operations: INSERT, DELETE
```

Comment:
This and the following tests are not necessary since the intent was just to show the issue would not occur in scenarios with just two ops.

Reply:
removed